File: dms_plus/dms_plus/report/pipeline_report/pipeline_report.js
// Copyright (c) 2025, dms_plus developers and contributors
// For license information, please see license.txt

let statuses = [];
frappe.query_reports["Pipeline Report"] = {
	filters: [
		{
			fieldname: "sales_person",
			label: __("Sales Person"),
			fieldtype: "Link",
			options: "Sales Person",
			reqd: 0
		},
		{
			fieldname: "quotation_status",
			label: __("Quotation Status"),
			fieldtype: "Select",
			options: [""],
			default: "",
			reqd: 0
		},
		{
			fieldname: "sales_order_status",
			label: __("Sales Order Status"),
			fieldtype: "Select",
			options: [""],
			default: "",
			reqd: 0
		},
		{
			fieldname: "item_group",
			label: __("Item Group"),
			fieldtype: "Link",
			options: "Item Group",
			reqd: 0
		},
		{
			fieldname: "product_manager",
			label: __("Product Manager"),
			fieldtype: "Link",
			options: "User",
			reqd: 0
		},
		{
			fieldname: "item",
			label: __("Item"),
			fieldtype: "Link",
			options: "Item",
			reqd: 0
		},
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
			reqd: 0
		},
		{
			fieldname: "territory",
			label: __("Territory"),
			fieldtype: "Link",
			options: "Territory",
			reqd: 0
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -3)
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today()
		}
	],

	onload: function (report) {
		report.page.set_title(__("Sales Pipeline Report"));
		report.trigger_refresh_on_load = true;

		report.page.add_inner_button(__("Export Report"), function () {
			frappe.query_report.export_report("CSV");
		});

		getSalesOrderStatus(report);
		getQuotationStatus(report);
	},

	after_datatable_render: function (datatable_wrapper) {
		$(datatable_wrapper).find('table').addClass('table-striped table-hover');
		statuses = ["Cancelled", "Draft", "To Deliver and Bill", "To Deliver", "To Bill", "Completed"];
		updatePipelineSummary(statuses);
		addReportStyling();

	},

	formatter: function (value, row, column, data, default_formatter) {

		if (column.fieldname === "so_status") {

			let status_color = "badge-secondary";

			switch (value) {
				case "Draft":
					status_color = "badge-info";
					break;
				case "To Deliver and Bill":
					status_color = "badge-warning";
					break;
				case "To Deliver":
					status_color = "badge-primary";
					break;
				case "To Bill":
					status_color = "badge-danger";
					break;
				case "Completed":
					status_color = "badge-success";
					break;
				case "Cancelled":
					status_color = "badge-secondary";
					break;
			}

			return `<span class="badge ${status_color}"
						style="
						padding: 4px 20px;
						border-radius: 5px;
						font-size:10px;
						align-items: center;
						justify-content: center;
						display: flex;">${value}</span>`;
		}
		if (column.fieldname === "pipeline_status" || column.fieldname === "quotation_status") {
			let color = "#999";
			let background = "#f0f0f0";

			switch (value) {
				case "Open":
					color = "#0c5aa0";
					background = "#cfe9f3";
					break;
				case "Draft":
					color = "#0c5aa0";
					background = "#cfe9f3";
					break;
				case "To Deliver and Bill":
					color = "#856404";
					background = "#fff3cd";
					break;
				case "To Deliver":
					color = "#0c5aa0";
					background = "#d1ecf1";
					break;
				case "To Bill":
					color = "#721c24";
					background = "#f8d7da";
					break;
				case "Completed":
					color = "#155724";
					background = "#d4edda";
					break;
				case "Cancelled":
					color = "#383d41";
					background = "#e2e3e5";
					break;
			}

			return `<span
						style="background-color: ${background};
						color: ${color};
						padding: 2px 5px;
						border-radius: 5px;
						font-size:10px;
						align-items: center;
						justify-content: center;
						display: flex;">${value}
					</span>`;
		}
		return default_formatter(value, row, column, data);
	}
};

function getSalesOrderStatus(report) {
	frappe.model.with_doctype("Sales Order", function () {

		let field = frappe.meta.get_docfield("Sales Order", "status", "Sales Order");
		if (field && field.options) {

			let options = field.options.split('\n').filter(opt => opt.trim() !== "");
			options.unshift("");
			let status_filter = report.get_filter('sales_order_status');
			status_filter.df.options = options;
			status_filter.refresh();

		} else {
			console.warn("Couldn't find status field or it has no options");
		}
	});
}

function getQuotationStatus(report) {
	frappe.model.with_doctype("Quotation", function () {

		let field = frappe.meta.get_docfield("Quotation", "status", "Quotation");
		if (field && field.options) {

			let options = field.options.split('\n').filter(opt => opt.trim() !== "");
			options.unshift("");
			let status_filter = report.get_filter('quotation_status');
			status_filter.df.options = options;
			// statuses = options;
			status_filter.refresh();
			return options;
		} else {
			console.warn("Couldn't find status field or it has no options");
		}
	});
}

function build_filters(filters_array) {
	let filters = {};

	if (filters_array && filters_array.length) {
		filters_array.forEach(function (f) {
			if (f.value) {
				filters[f.key] = f.value;
			}
		});
	}

	return filters;
}

async function updatePipelineSummary(statuses_list) {
	$('#pipeline-summary-custom').remove();

	let data = [];
	if (frappe.query_report && frappe.query_report.datatable) {
		data = frappe.query_report.datatable.datamanager.data || [];
	}

	if (!data || data.length === 0 || !statuses_list || statuses_list.length === 0) return;

	const summary = {};
	// Pipeline statuses to summarize
	// Statuses come from the Quotation Status and Sales Order Status filters
	// Example statuses: ["Draft", "Open", "To Deliver", "To Bill", "Completed", "Cancelled"]
	// statuses_list = statuses_list.filter(status => status && status.trim() !== "");

	statuses_list.forEach(status => {
		summary[status] = data.filter(row =>
			row.pipeline_status === status || row.status === status
		).length;
	});

	console.log("✅ cleaned statuses_list:", statuses_list);
	console.log("✅ pipeline summary data:", summary);
	console.log("✅ full data:", data);
	let summary_html = `

		<div id="pipeline-summary-custom" class="frappe-control frappe-card-section"
		 		style="
					margin-bottom: 20px;
					padding: 20px;
					width: 100%;
					border-radius: 8px;
				">
			<div class="form-section">
				<h6 class="section-head" style="margin-bottom: 15px; display: flex; align-items: center; justify-content: center;">
					<i class="fa fa-line-chart" style="margin-right: 8px;"></i>
					<span>Quotation Status Summary</span>
				</h6>
				<div class="form-grid-row" style="display: grid; grid-template-columns: repeat(auto-fit, minmax(150px, 1fr)); gap: 12px;">
	`;

	statuses_list.forEach(status => {
		const count = summary[status] || 0;
		summary_html += `
			<div style="
				background: rgba(255, 255, 255, 0.15);
				padding: 18px;
				border-radius: 8px;
				text-align: center;
				border: 1px solid rgba(255, 255, 255, 0.3);
				transition: all 0.3s;
			">
				<div style="font-size: 12px; opacity: 0.95; margin-bottom: 10px; letter-spacing: 0.5px; font-weight: 600;">${status}</div>
				<div style="font-size: 32px; font-weight: 900;">${count}</div>
			</div>
		`;
	});

	summary_html += `
			</div>
		</div>
	`;

	const reportResult = $('div.result');
	console.log("✅ Inserting pipeline summary into report...", reportResult.length);
	if (reportResult.length) {
		console.log("✅ Prepending pipeline summary");
		reportResult.prepend(summary_html);
	} else {
		console.log("✅ Prepending pipeline summary 2");
		$('div.frappe-control').last().after(summary_html);
	}
}

function addReportStyling() {
	console.log("Adding custom styles to Pipeline Report...");
	const style = `
		.datatable table thead {
			background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
			color: white;
		}

		.datatable table th {
			color: white;
			font-weight: 700;
			text-transform: uppercase;
			letter-spacing: 0.5px;
		}

		.datatable table tbody tr:hover {
			background: #f5f5f5 !important;
		}
	`;

	if ($('#pipeline-report-styles').length === 0) {
		$('head').append(`<style id="pipeline-report-styles">${style}</style>`);
	}
}

$(document).on('frappe-query-report-ready', function () {
	setTimeout(function () {
		console.log("Updating pipeline summary on report ready...");
		updatePipelineSummary(statuses);
	}, 500);
});

frappe.query_report.refresh_on_filter_change = true;

File: dms_plus/dms_plus/report/pipeline_report/pipeline_report.py
# # Copyright (c) 2025, dms_plus developers and contributors
# # For license information, please see license.txt
import frappe
from frappe import _

def execute(filters: dict | None = None):
    """Main entry point for Pipeline Report"""
    columns = get_columns()
    data = get_quotation_data(filters)
    report_summary = generate_pipeline_summary(data)

    return columns, data, None, None, report_summary

def get_columns() -> list[dict]:
    """Define report columns"""
    return [
        {
            "label": _("Quotation"),
            "fieldname": "quotation_name",
            "fieldtype": "Link",
            "options": "Quotation",
            "width": 200
        },
        {
            "label": _("Customer"),
            "fieldname": "party_name",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 200
        },
        {
            "label": _("Item"),
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 150
        },
        {
            "label": _("Item Group"),
            "fieldname": "item_group",
            "fieldtype": "Link",
            "options": "Item Group",
            "width": 150
        },
        {
            "label": _("Product Manager"),
            "fieldname": "product_manager",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "label": _("Order Date"),
            "fieldname": "transaction_date",
            "fieldtype": "Date",
            "width": 200
        },
        {
            "label": _("Delivery Date"),
            "fieldname": "delivery_date",
            "fieldtype": "Date",
            "width": 200
        },
        {
            "label": _("Qty"),
            "fieldname": "qty",
            "fieldtype": "Int",
            "width": 80
        },
        {
            "label": _("Amount"),
            "fieldname": "amount",
            "fieldtype": "Currency",
            "width": 180
        },
        {
            "label": _("Quotation Status"),
            "fieldname": "quotation_status",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Related Sales Order"),
            "fieldname": "related_sales_order",
            "fieldtype": "Link",
            "options": "Sales Order",
            "width": 150
        },
        {
            "label": _("SO Status"),
            "fieldname": "so_status",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("% Delivered"),
            "fieldname": "per_delivered",
            "fieldtype": "Percent",
            "width": 120
        },
        {
            "label": _("% Billed"),
            "fieldname": "per_billed",
            "fieldtype": "Percent",
            "width": 120
        },
        {
            "label": _("Pipeline Status"),
            "fieldname": "pipeline_status",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Territory"),
            "fieldname": "warehouse",
            "fieldtype": "Link",
            "options": "Territory",
            "width": 120
        },
        {
            "label": _("Sales Person"),
            "fieldname": "sales_person",
            "fieldtype": "Data",
            "width": 150
        }
    ]

def build_filters(filters: dict) -> tuple[str, dict]:
    """Build SQL WHERE conditions based on filters"""
    where_conditions = []
    params = {}

    print("\nfilters.get quotation_status",filters.get("quotation_status"))
    if filters.get("quotation_status"):
        quotation_status = filters.get("quotation_status")
        where_conditions.append("q.status = %(quotation_status)s")
        params["quotation_status"] = filters.get("quotation_status")
    if filters.get("sales_order_status"):
        where_conditions.append("so.status = %(so_status)s")
        params["so_status"] = filters.get("sales_order_status")
    if filters.get("item_group"):
        where_conditions.append("qi.item_group = %(item_group)s")
        params["item_group"] = filters.get("item_group")
    if filters.get("item"):
        where_conditions.append("qi.item_code = %(item)s")
        params["item"] = filters.get("item")
    if filters.get("product_manager"):
        where_conditions.append("q.owner = %(product_manager)s")
        params["product_manager"] = filters.get("product_manager")
    if filters.get("customer"):
        where_conditions.append("q.party_name = %(customer)s")
        params["customer"] = filters.get("customer")
    if filters.get("territory"):
        where_conditions.append("q.territory = %(territory)s")
        params["territory"] = filters.get("territory")
    if filters.get("from_date"):
        where_conditions.append("q.transaction_date >= %(from_date)s")
        params["from_date"] = filters.get("from_date")
    if filters.get("to_date"):
        where_conditions.append("q.transaction_date <= %(to_date)s")
        params["to_date"] = filters.get("to_date")

    where_clause = " AND ".join(where_conditions) if where_conditions else "1=1"
    return where_clause, params

def get_quotation_data(filters: dict) -> list:
    """Fetch Quotations and their related Sales Orders"""
    filters = filters or {}
    print("\nfilters in get_quotation_data",filters )
    # Build WHERE conditions for Quotations
    where_clause, params = build_filters(filters)
    print("\nwhere_clause",where_clause )
    print("\nparams",params )
    query = f"""
        SELECT DISTINCT
            q.name as quotation_name,
            q.party_name,
            qi.item_code,
            qi.item_group,
            q.owner as product_manager,
            q.transaction_date,
            q.valid_till as delivery_date,
            qi.qty,
            qi.amount,
            qi.warehouse,
            q.status as quotation_status,
            q.territory,
            qt.sales_person
        FROM `tabQuotation` q
        LEFT JOIN `tabQuotation Item` qi ON qi.parent = q.name
        LEFT JOIN `tabSales Team` qt ON qt.parent = q.name
        WHERE {where_clause}
        ORDER BY q.transaction_date DESC
    """

    try:
        results = frappe.db.sql(query, params, as_dict=True)

        # أضف Related Sales Order والـ Pipeline Status لكل Quotation
        for row in results:
            # print("\n\nrow before SO fetch",row )
            sales_order_data = get_sales_order_for_quotation(row.get('quotation_name'), row.get('party_name'))
            # print("sales_order_data",sales_order_data )
            row['related_sales_order'] = sales_order_data.get('sales_order_name', 'No SO Found')
            row['so_status'] = sales_order_data.get('status', 'No SO Found')
            row['sales_person'] = ', '.join(get_sales_persons(row['related_sales_order'])) if row['related_sales_order'] != 'No SO Found' else ''
            row['per_delivered'] = sales_order_data.get('per_delivered', 0)
            row['per_billed'] = sales_order_data.get('per_billed', 0)

            if sales_order_data.get('sales_order_name'):
                row['pipeline_status'] = calculate_pipeline_status(sales_order_data)
            else:
                row['pipeline_status'] = 'Not Converted'
        # print("\n\ntotal quotation",results )
        # print("where_clause",where_clause )
        return results
    except Exception as e:
        frappe.msgprint(f"Error fetching data: {str(e)}")
        return []


def get_sales_order_for_quotation(quotation_name: str, customer:str) -> dict:
    """
    Get the latest Sales Order related to a Quotation
    based on the prevdoc_docname field in Sales Order Item.
    """
    try:
        sales_orders = frappe.db.sql(
            """
            SELECT DISTINCT
                soi.parent AS sales_order_name,
                so.status,
                so.per_delivered,
                so.per_billed,
                so.transaction_date
            FROM `tabSales Order Item` soi
            LEFT JOIN `tabSales Order` so ON so.name = soi.parent
            WHERE soi.prevdoc_docname = %s
            And so.customer = %s
            AND so.status != 'Cancelled'
            ORDER BY so.transaction_date DESC
            LIMIT 1
            """,
            (quotation_name, customer),
            as_dict=True
        )

        return sales_orders[0] if sales_orders else {}

    except Exception as e:
        frappe.log_error(f"Error getting sales order: {str(e)}")
        return {}



def calculate_pipeline_status(row: dict) -> str:
    """
    calculate the pipeline status based on delivery and billing percentages.
    status can be:
    - Cancelled
    - Draft
    - Completed
    - To Bill
    - To Deliver
    - To Deliver and Bill
    """
    status = row.get('status', 'Draft')
    per_delivered = row.get('per_delivered', 0) or 0
    per_billed = row.get('per_billed', 0) or 0

    # تحويل النسب إلى أرقام عادية
    if isinstance(per_delivered, str):
        per_delivered = float(per_delivered.replace('%', '') or 0)
    if isinstance(per_billed, str):
        per_billed = float(per_billed.replace('%', '') or 0)

    # منطق تحديد الحالة
    if status == "Cancelled":
        return "Cancelled"
    elif status == "Draft":
        return "Draft"
    elif per_delivered == 100 and per_billed == 100:
        return "Completed"
    elif per_delivered == 100 and per_billed < 100:
        return "To Bill"
    elif per_delivered > 0 and per_delivered < 100 and per_billed == 0:
        return "To Deliver"
    elif per_delivered > 0 and per_delivered < 100 and per_billed > 0:
        return "To Deliver and Bill"
    elif per_delivered == 0 and per_billed == 0:
        return "To Deliver and Bill"
    else:
        return "To Deliver and Bill"

def get_sales_persons(sales_order_name) -> list:
    """Fetch all Sales Persons for filter options."""
    sales_persons = frappe.db.get_all("Sales Team", fields=["*"], filters={
        "parenttype": "Sales Order",
        "parent": sales_order_name
    })
    return [sp.sales_person for sp in sales_persons]

def generate_pipeline_summary(data: list) -> list:
    """Generate report_summary (shown above report table)"""
    summary = {}
    for row in data:
        status = row.get('pipeline_status', 'Unknown')
        summary[status] = summary.get(status, 0) + 1

    report_summary = []
    for status, count in summary.items():
        report_summary.append({
            "value": count,
            "label": status,
            "datatype": "Int"
        })

    return report_summary
